### ISSUE TITLE
Sketch: Faces with normals in opposite directions aren't combined into a single Face

### ISSUE THREAD

**Issue by gumyr** [gumyr/build123d#267]
Consider the following example:
```py
from build123d import *
from ocp_vscode import *

f1 = Face.make_from_wires(Wire.make_polygon([(1, 0), (2, 0), (2, 1), (1, 1), (1, 0)]))
f2 = Face.make_from_wires(Wire.make_polygon([(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)]))
f3 = f1.fuse(f2)
print(f3.show_topology("Face"))
show_all()
```
![image](https://github.com/gumyr/build123d/assets/85638518/5168013f-38a9-4399-9e28-c2ed37116afb)
```
Compound at 0x7fccaa533bf0, Center(1.0, 0.5, 0.0)
├── Face at 0x7fccaf5b4070, Center(1.5, 0.5, 0.0)
└── Face at 0x7fccaa533bb0, Center(0.5, 0.5, 0.0)
```
When using `add` to add new faces to a sketch, the normals must all face in the positive Z direction.

**Comment by gumyr** [comment]
More data:
```py
with BuildSketch() as t1:
    Rectangle(1, 1, align=Align.MIN)
    with Locations((1, 0)):
        Rectangle(1, 1, align=Align.MIN)

f1 = Face.make_from_wires(Wire.make_polygon([(1, 0), (2, 0), (2, 1), (1, 1), (1, 0)]))
f2 = Face.make_from_wires(Wire.make_polygon([(0, 0), (1, 0), (1, 1), (0, 1), (0, 0)]))
f3 = f1.fuse(f2)
```
t1 is a single Face but f3 is a Compound of two faces.  In both cases the normals are in the +ve Z direction.

**Comment by bernhard-42** [comment]
I looked into it and I think one needs to reverse the wires of the face in order to flip it completely:

Turns out that it seems sufficient to flip the `LINE` edges:

```python
def reverse_wire(wire):
    wire_builder = BRepBuilderAPI_MakeWire()
    combined_edges = TopTools_ListOfShape()
    for edge in reversed(wire.edges()):
        v = edge.vertices()
        t = edge.geom_type()
        if t == "LINE":
            rev_edge = BRepBuilderAPI_MakeEdge(v[1].wrapped, v[0].wrapped).Edge()
        else:
            rev_edge = edge.wrapped
        combined_edges.Append(rev_edge)
    wire_builder.Add(combined_edges)
    wire_builder.Build()
    return Wire(wire_builder.Wire())
```

It also seems sufficient to reverse the `outer_wire`:

```python
def flip_face(face):
    rev_outer_wire = []
    inner_wires = []
    for wire in face.wires():
        if wire == face.outer_wire():
            rev_outer_wire = reverse_wire(wire)
        else:
            inner_wires.append(wire)

    return Face.make_from_wires(rev_outer_wire, inner_wires)
```

Example:
```python
f1 = Face.make_from_wires(
    Wire.make_polygon([(1, 0), (1.5, 0.5), (1, 2), (3, 1), (2, 0), (1, 0)])
)
f1 = (
    fillet(f1.vertices()[2:4], 0.2)
    - Pos(1.8, 1.2, 0) * Rectangle(0.2, 0.4)
    - Pos(2, 0.5, 0) * Circle(0.2)
).faces()[0]

f2 = Face.make_from_wires(
    Wire.make_polygon([(1, 0), (1.5, -1), (2, -1), (2, 0), (1, 0)])
)
show(f1, f2)
```

<img width="194" alt="image" src="https://github.com/gumyr/build123d/assets/954408/947246bd-bb2c-4cea-8efc-5957e31a4526">

```python
show(flip_face(f1), f2)
```

<img width="194" alt="image" src="https://github.com/gumyr/build123d/assets/954408/c3977530-421d-4745-a082-8dbc586bb0ee">
Now let's fuse it:

```python
f3 = flip_face(f1) + f2

show(f3)
print(f3.show_topology("Face"))

# +Compound at 0x169e67370, Center(1.931176925864295, 0.40255330702035463, 0.0)
# └── Face at 0x157f784b0, Center(1.931176925864295, 0.40255330702035463, 0.0)
```

<img width="194" alt="image" src="https://github.com/gumyr/build123d/assets/954408/2a41f9ea-7741-4a37-89e6-589ce28a39d9">

Not sure it would work for all wires with all line types, but thought to share

**Comment by bernhard-42** [comment]
> More data:
> 
> ```python
> with BuildSketch() as t1:
>     Rectangle(1, 1, align=Align.MIN)
>     with Locations((1, 0)):
>         Rectangle(1, 1, align=Align.MIN)
> 
> f1 = Face.make_from_wires(Wire.make_polygon([(1, 0), (2, 0), (2, 1), (1, 1), (1, 0)]))
> f2 = Face.make_from_wires(Wire.make_polygon([(0, 0), (1, 0), (1, 1), (0, 1), (0, 0)]))
> f3 = f1.fuse(f2)
> ```
> 
> t1 is a single Face but f3 is a Compound of two faces. In both cases the normals are in the +ve Z direction.

Try `f1.fuse(f2).clean()` or `f1 + f2`
This leads to the expected result

and both the builder and the algebra mode apply `clean`.after `fuse`. Hence the difference as far as I can see

**Comment by bernhard-42** [comment]
Assume the above `flip_face` would work in general, then the part missing in `fuse` would be to check whether all `to_fuse` faces are `PLANE` and then `flip_face(face)` of  all `face` with `face.normal_at() == -self.normal_at()`.

**Comment by bernhard-42** [comment]
In any case, this is just a sketch of a solution for this problem. But it does look promising to me

**Comment by gumyr** [comment]
Thanks for looking into this.  There is an OCP method `Complemented` that does the flip of a face - whith is what is used in the `Face.__neg__` operation.  Here are some examples:
```py
from build123d import *
from ocp_vscode import *

test = Face.make_from_wires(Wire.make_rect(1, 1), [Wire.make_circle(0.2)])
print(test.wrapped.Orientation())
print(test.normal_at())
test_reversed = Face(test.wrapped.Complemented().Moved(Location((1.5, 0, 0)).wrapped))
print(test_reversed.wrapped.Orientation())
print(test_reversed.normal_at())
test_neg = -test.moved(Location((0, 1.5, 0)))
print(test_neg.wrapped.Orientation())
print(test_neg.normal_at())

show_all()
```
```
TopAbs_Orientation.TopAbs_FORWARD
Vector: (-0.0, 0.0, -1.0)
TopAbs_Orientation.TopAbs_REVERSED
Vector: (0.0, -0.0, 1.0)
TopAbs_Orientation.TopAbs_REVERSED
Vector: (0.0, -0.0, 1.0)
```
![image](https://github.com/gumyr/build123d/assets/85638518/3c8f1217-7b5f-40f2-be3c-7ef164e8f8db)

The fix should be just finding the flipped faces and using `-` to flip them back before adding them to the sketch.

**Comment by gumyr** [comment]
Fixed with commit 6aa434042b019dc2bb2d0daa6a0ee58d4008f5a7  The change in the `build_common.py` was quite small:
```py
-                        reoriented_face = plane.to_local_coords(face)
-                        aligned.append(
-                            reoriented_face.moved(
-                                Location((0, 0, -reoriented_face.center().Z))
-                            )
-                        )
-                    else:
+                        face: Face = plane.to_local_coords(face)
+                        face.move(Location((0, 0, -face.center().Z)))
+                    if face.normal_at().Z > 0:  # Flip the face if up-side-down
                         aligned.append(face)
+                    else:
+                        aligned.append(-face)
```
Now:
```py
f1 = Face.make_from_wires(
    Wire.make_polygon([(1, 0), (1.5, 0.5), (1, 2), (3, 1), (2, 0), (1, 0)])
)
f1 = (
    fillet(f1.vertices()[2:4], 0.2)
    - Pos(1.8, 1.2, 0) * Rectangle(0.2, 0.4)
    - Pos(2, 0.5, 0) * Circle(0.2)
).faces()[0]
print(f"{f1.normal_at()=}")

f2 = Face.make_from_wires(
    Wire.make_polygon([(1, 0), (1.5, -1), (2, -1), (2, 0), (1, 0)])
)
print(f"{f2.normal_at()=}")
with BuildSketch() as flip_test:
    add(f1)
    add(f2)

show(f1, f2, flip_test.sketch.moved(Location((-2, 0, 0))))
```
results in:
![image](https://github.com/gumyr/build123d/assets/85638518/f3ce61f9-b87b-4a71-9142-760d485854f7)
